receiver/statsdreceiver/transport/udp_server.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net"
+	"strings"
+	"sync"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver/protocol"
+)
+
+type udpServer struct {
+	wg         sync.WaitGroup
+	packetConn net.PacketConn
+	reporter   Reporter
+}
+
+var _ (Server) = (*udpServer)(nil)
+
+// NewUDPServer creates a transport.Server using UDP as its transport.
+func NewUDPServer(addr string) (Server, error) {
+	packetConn, err := net.ListenPacket("udp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	u := udpServer{
+		packetConn: packetConn,
+	}
+	return &u, nil
+}
+
+func (u *udpServer) ListenAndServe(
+	parser protocol.Parser,
+	nextConsumer consumer.MetricsConsumerOld,
+	reporter Reporter,
+) error {
+	if parser == nil || nextConsumer == nil || reporter == nil {
+		return errNilListenAndServeParameters
+	}
+
+	u.reporter = reporter
+
+	buf := make([]byte, 65527) // max size for udp packet body (assuming ipv6)
+	for {
+		n, _, err := u.packetConn.ReadFrom(buf)
+		if n > 0 {
+			u.wg.Add(1)
+			bufCopy := make([]byte, n)
+			copy(bufCopy, buf)
+			go func() {
```

Comment:
Do the other receivers have this pattern of spawning a goroutine? I'd hope the reporter handles any needed handoff to not worry about it here.

Reply:
I followed the `carbonreceiver` UDP server implementation which also does this goroutine spawning.

Reply:
Ok it could mean that `nextConsumer.ConsumeData` can block so we need this. Will be nice to confirm that since otherwise it does seem to introduce reasonable overhead.

Reply:
Yes, in principle there is no telling what next consumer is going to do. The collector in general tries to avoid have clients waiting (one of the points of the collector is to avoid burden on apps).

Reply:
I think that makes sense - it seems like this could be abstracted into the collector though as a framework. For example, the call to `nextConsumer.ConsumeMetrics` instead of directly calling the consumer is first calling into framework code that e.g., spawns a goroutine, which could simply the consumer implementations. Just an idea though.

Reply:
I'm back and need to clarify this point :) /cc @wangzlei 

We may use the collector in a synchronous context, for example if it needs to quickly export metrics before a serverless runtime shuts down. So I feel like unless an "async" processor, such as a batching processor, is added, it actually should allow clients to wait or else this is just not supported at all. This should definitely not be a recommended configuration, but I think there are use cases for synchronous export - if receivers spawn goroutines though that becomes impossible IIUC.

Reply:
I believe the reason that UDP is the most common transport for Statsd is that it is non-blocking. Therefore, I think it's OK to block the receiver. If concurrency is desired, then I would expect to add more goroutines reading and processing packets synchronously.

Reply:
Ok cool, so it sounds like I should remove this goroutine spawning and we can add concurrency in behind further configuration options later?